core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -264,12 +264,37 @@ class TaskMetrics private[spark] () extends Serializable {
   /**
    * External accumulators registered with this task.
    */
-  @transient private[spark] lazy val _externalAccums = new CopyOnWriteArrayList[AccumulatorV2[_, _]]
+  @transient private[spark] lazy val _externalAccums = new ArrayBuffer[AccumulatorV2[_, _]]
 
-  private[spark] def externalAccums = _externalAccums.asScala
+  private[spark] def externalAccums = withReadLock {
+    _externalAccums
+  }
```

Comment:
By returning `_externalAccums`, we are allowing it to escape the read lock and get used directly in non-MT safe contexts (like `accumulators`).

Reply:
How do we handle such cases in general? Seems it's impossible to avoid the misuse completely from the caller.

Reply:
Or maybe we should return an immutable copy of `_externalAccums`. This brings extra overhead though.

Reply:
As an example, `ShuffleStatus.withReadLock` has one such formulation on how to do it in a thread safe manner.
Given `externalAccums`  is `private[spark]`, should be possible to adapt something like that ... though this might be slightly cumbersome.

My preference would be to:
a) Replace existing high frequency uses of `externalAccums` within `TaskMetrics` by surrounding with `withReadLock` in `core` - and not use  `externalAccums` directly (for ex `accumulators`, `Task.collectAccumulatorUpdates`).
b) For all external use (like from `sql`), return a `.toList` within the read lock:
```
private[spark] def externalAccums = withReadLock {
    _externalAccums.toList
  }
```

Thoughts ?

Reply:
Shall we simply change the return type of `def externalAccums` to be `Seq[AccumulatorV2[_, _]]`?

Reply:
Looking at this more, I assumed `accumulators` was used a bit more than that it currently is (which is the one place where I thought something like this would help: `private[spark] def accumulators(): Seq[AccumulatorV2[_, _]] = withReadLock { internalAccums ++ externalAccums }`) - looks like it is only in heartbeats - so not that common as well.

Given all other usages of `externalAccums` and `accumulators` is part of task end - conversion from `ArrayBuffer` to an immutable list should be fine.

So let us simply make it:
```
private[spark] def externalAccums = withReadLock {
    _externalAccums.toList
}
```
Note, we have to explicitly create an (immutable) copy within the read lock (keeping [this comment](https://github.com/apache/spark/pull/47197#discussion_r1665413890) in mind as well)